### PR TITLE
Replace stubs in ndimage by Jitify

### PR DESCRIPTION
**Pull request by leofang**
Follow-up of #4228. We no longer need customized C++ std code now that we have Jitify.

TODO:
- [x] Test HIP

### CODE REVIEW

**Comment by leofang** [comment]
Looks like ROCm/HIP can handle `type_traits` as expected, but there're some issues that many tests fail (#4268), which is irrelevant of the present PR.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Converted to draft and check if I can make some improvements to upstream (Jitify).

**Comment by leofang** [comment]
> Converted to draft and check if I can make some improvements to upstream (Jitify): [NVIDIA/jitify#82](https://github.com/NVIDIA/jitify/pull/82)

Blocked by that and also #4277. After that the JIT overhead will be very small.

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/cuda/jitify.pyx [hunk 1]
```diff
@@ -36,8 +38,18 @@ def get_build_version():
     return jitify_ver.decode()
 
 
-# cache all headers; this is intialized with built-in JIT-safe headers
-cdef cpp_map[cpp_str, cpp_str] cupy_headers = get_jitsafe_headers_map()
+# We cache headers found by Jitify. This is initialized with a few built-in
+# JIT-safe headers, and expands as needed to help reduce compile time.
+cdef cpp_map[cpp_str, cpp_str] cupy_headers
```
Comment (leofang) [inline cupy/cuda/jitify.pyx:43]: It turns out the previous improvement (#4277) was too aggressive that could lead to slowdown in compile time, so I revert the initialization strategy to what was introduced originally (#4228).

**Comment by leofang** [comment]
I did quite a bit tests in order to evaluate the impact of Jitify overhead. It turns out that we need #4277 + the patch applied to `cupy/cuda/jitify,pyx` (included in this PR) in order to minimize the compile time, see below:

time (roughly) | before | after (no change to `jitify.pyx`) | after (with `jitify.pyx` fixed)
--------|-------|------|-------
`ndimage_tests/test_filters.py` | 10 min | 20 min | 16 min 
`ndimage_tests/` | 30 min | 45 min | 37.5 min

Note that `test_filters.py` was picked because it's affected the most, as can be seen from the increased timing. I think this PR made the overhead acceptable (there are ~5.7k tests in `test_filters.py`, so the averaged overhead, assuming each test compiles 1 new kernel, is (6*60)/5700 = 63ms per test).

**Comment by leofang** [comment]
@emcastillo This is ready to review.

**Review by toslunar — APPROVED** [review]
LGTM

**Comment by toslunar** [comment]
pfnCI, test this please.

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Is the CI timeout shortened? In `cupy-py3` all ndimage tests actually passed, but the CI was terminated. Let me retry...

Jenkins, test this please

**Comment by leofang** [comment]
Thanks, @toslunar @emcastillo!

**Pull request merged**